### PR TITLE
Optimize the Dockerfile

### PR DESCRIPTION
**Pull request by eriktews**
This optimizes the Dockerfile in a way that a lot of caching for subsequent builds can be used. Also the installer is split into multiple smaller parts that can be mostly shared with the Dockerfile.

### CODE REVIEW

**Comment by eriktews** [comment]
This should also fix #304 

There are maybe a few more optimizations, but I think the new version is on a good level.

File: install-system.sh [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+        flash=true
+        echo Installing Adobe Flash Plugin
+    else
+        echo Unrecognized response, exiting
+        exit 1
+    fi
+fi
+
+if [ "$flash" = true ]; then
+    sudo sh -c 'echo "deb http://archive.canonical.com/ubuntu/ trusty partner" >> /etc/apt/sources.list.d/canonical_partner.list'
```
Comment (englehardt) [inline install-system.sh]: Let's update this to `sudo add-apt-repository "deb http://archive.canonical.com/ $(lsb_release -sc) partner"` so we use the correct archive for whatever version of the OS the user is using.

File: install-node.sh [hunk 2]
```diff
@@ -0,0 +1,6 @@
+#!/bin/bash
```
Comment (englehardt) [inline install-node.sh]: From what I can tell, https://github.com/mozilla/OpenWPM/blob/master/install-dev.sh isn't needed anymore. `jpm` is the build tool for the old add-on sdk extensions, and we install node separately in this file. If you agree, do you mind to remove `install-dev.sh` and `requirements-dev.txt` as part of this PR? The latter only contains `autopep8`, which someone can install if they wish -- it's not used by anything in the repo.

File: install-system.sh [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+# tail -n 1 | sed -e 's/.*releases\///g' | cut -d '/' -f1)"
+#
+# wget https://ftp.mozilla.org/pub/firefox/releases/${firefox_version}/linux-$(uname -m)/en-US/firefox-${firefox_version}.tar.bz2
+# tar jxf firefox-${firefox_version}.tar.bz2
+
+wget https://index.taskcluster.net/v1/task/gecko.v2.mozilla-release.latest.firefox.linux64-add-on-devel/artifacts/public/build/target.tar.bz2
+tar jxf target.tar.bz2
+rm -rf firefox-bin
+mv firefox firefox-bin
+# rm firefox-${firefox_version}.tar.bz2
```
Comment (englehardt) [inline install-system.sh]: We can remove this comment.

File: README.md [hunk 4]
```diff
@@ -514,31 +514,38 @@ Docker service.
 __Step 2:__ to build the image, run the following command from a terminal
 within the root OpenWPM directory:
 
+```
     docker build -f Dockerfile -t openwpm .
-
-After building the above, you may optionally build a docker image for OpenWPM development:
```
[the remark this answers is not in the document]
Comment (eriktews) [reply README.md]: Right now, there is no dedicated -dev Dockerfile anymore, but I will add that later.

File: install-system.sh (hunk 1 above)
Comment (eriktews) [reply install-system.sh]: Done

File: install-node.sh (hunk 2 above)
Comment (eriktews) [reply install-node.sh]: This is correct, they can be removed.

File: install-system.sh (hunk 3 above)
Comment (eriktews) [reply install-system.sh]: Done, and I also removed the other comments related to firefox-${firefox_version}.tar.bz2 as well

File: README.md [hunk 5]
```diff
@@ -514,31 +514,38 @@ Docker service.
...
+
+Instead of _python_, _python3_ can be used here as well.
 
 This command uses _bind-mounts_ to share scripts and output between the
 container and host, as explained below (note the paths in the command assume
 it's being run from the root OpenWPM directory):
 
 - `run` starts the `openwpm` container and executes the
-    `python /opt/OpenWPM/demo.py` command.
+    `python /opt/OpenWPM/demo.py"` command.
```
[the remark this answers is not in the document]
Comment (eriktews) [reply README.md]: Fixed

File: Dockerfile [hunk 6]
```diff
@@ -1,12 +1,62 @@
...
+# Technically, the automation/Extension/firefox directory could be skipped
+# here, but there is no nice way to do that with the Docker COPY command
+COPY . .
+
+# Optionally create an OpenWPM user. This is not strictly required since it is
+# possible to run everything as root as well.
+RUN adduser --disabled-password --gecos "OpenWPM"  openwpm
+
+# Alternatively, python3 could be used here
 CMD python demo.py
```
[the remark this answers is not in the document]
Comment (eriktews) [reply Dockerfile]: No, it's not executed during the build. Just when someone starts an container with that image using the default arguments, demo.py is executed as the default script.

File: README.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (englehardt) [reply README.md]: It's still in the tree of your branch: https://github.com/mozilla/OpenWPM/blob/cef2ef77c843ecf4350bb93205f277f04d718762/Dockerfile-dev

File: Dockerfile (hunk 6 above)
[the remark this answers is not in the document]
Comment (englehardt) [reply Dockerfile]: Ah I see. Thanks for the explanation.

File: Dockerfile [hunk 7]
```diff
@@ -1,12 +1,62 @@
-FROM ubuntu:18.04
+# Two stage Dockerfile to build OpenWPM
+# Stage 1 builds the extension, stage 2 builds the main OpenWPM image
+
+FROM node:10 as extension
+WORKDIR /usr/src/app
+
+# Copy the package.json and package-lock.json first and install all the
+# packages. Should the source code of the extension be changed without
+# altering the list of required packages, this stop can be cached by Docker.
```
[the remark this answers is not in the document]
Comment (eriktews) [reply Dockerfile]: Done

File: Dockerfile [hunk 8]
```diff
@@ -1,12 +1,62 @@
...
+COPY automation/Extension/firefox/. ./
+RUN npm run build
 
-WORKDIR /app
+
+# Stage 2, build the main OpenWPM image
+FROM ubuntu:18.04
+
+WORKDIR /opt/OpenWPM
+# This is just a performance evaluation and can be skipped by none us based
```
[the remark this answers is not in the document]
Comment (eriktews) [reply Dockerfile]: Done

File: Dockerfile [hunk 9]
```diff
@@ -1,12 +1,62 @@
...
+# Install the Ubuntu packages as well as firefox and the geckodriver first
+COPY ./install-system.sh .
+RUN ./install-system.sh --no-flash
+
+# Instead of running install-pip-and-packages.sh, the packages are installed
+# manually using pip and pip3 so that python2 and python3 are supported in the
+# final image.
+RUN apt-get -y install python-pip python3-pip
+
+# For some reasons, python3-publicsuffix doesn't work with pip3 at the moent,
```
[the remark this answers is not in the document]
Comment (eriktews) [reply Dockerfile]: Done

File: Dockerfile [hunk 10]
```diff
@@ -1,12 +1,62 @@
...
+# Node is not required, the extension is build in the first build stage so
+# there is no noeed to run install-node.sh and build-extension.sh
+
+# Technically, the automation/Extension/firefox directory could be skipped
+# here, but there is no nice way to do that with the Docker COPY command
+COPY . .
+
+# Optionally create an OpenWPM user. This is not strictly required since it is
+# possible to run everything as root as well.
+RUN adduser --disabled-password --gecos "OpenWPM"  openwpm
```
[the remark this answers is not in the document]
Comment (eriktews) [reply Dockerfile]: Running jobs with X11 as the user openwpm didn't work very well on my system. So I mentioned that the user openwpm exists in the README, but recommended it only for headless crawls.

When this should be extended, I would recommend to add a demo-headless.py.

[the remark this answers is not in the document]
Comment (englehardt) [reply Dockerfile]: That works for me, thanks!

File: README.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (englehardt) [reply README.md]: We decided to remove Dockerfile-dev since it doesn't contain anything more than the normal Dockerfile with the current refactor.

**Comment by motin** [comment]
It seems that this PR also wiped out install-dev.txt which is not related to Docker, and is still mentioned in https://github.com/mozilla/OpenWPM#installation

**Pull request merged**